### PR TITLE
JsonApiListener documentation

### PR DESCRIPTION
**Pull request by bravo-kernel**
Concept, please shoot.

### CODE REVIEW

**Comment by josegonzalez** [comment]
Love it!

**Comment by bravo-kernel** [comment]
Just tested the instructions with a virgin install and works as expected.

**Comment by bravo-kernel** [comment]
For those following this... I will create one additional PR, after that a release can be made as far as I'm concerned.

File: docs/listeners/jsonapi.rst [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+  {
+    "jsonapi": {
+      "version": "1.0"
+    }
+  }  
+  
+meta
+^^^^
+
+Pass this **mixed** option (default: false) an array or hash will make the listener add the
```
Comment (jippi) [inline docs/listeners/jsonapi.rst]: `array or object` ? in what cases does it do what?

**Comment by jippi** [comment]
LGTM

**Comment by bravo-kernel** [comment]
Thanks for reviewing (both). Two more things on my list before this can be merged:

- I will add a section about the underlying Schema logic and possibility to create custom schemas
- I will review the ``meta`` option. Seems like it could be just an ``array``, good point.

**Comment by bravo-kernel** [comment]
Added a adding section to the bottom of the page describing the use of ``Schema`` files which completes the docs IMO.

Mr. @davidyell, would you please proofread, I don't want to ruin all your hard work by breaking consistency.

**Comment by bravo-kernel** [comment]
Gents, once #471 and #472 get merged my task list will be complete for what I consider to be a pretty solid starting point for a first release. 

It might be worth waiting for some updates mentioned by @ceeram but that's not up to me, I will leave releasing to the big boys.

Regards ☺️

File: docs/listeners/jsonapi.rst [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+      parent::beforeFilter();
+      $this->Crud->addListener('Crud.JsonApi');
+    }
+  }
+
+
+CakeRequest detector
+--------------------
+
+The JsonApi Listener adds the ``is(jsonapi)`` request detector
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: Should this be `is('jsonapi')` as a string?

File: docs/listeners/jsonapi.rst [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+
+  {
+    "errors": [
+      {
+        "code": "501",
+        "title": "Not Implemented"
+      }
+    ],
+    "debug": {
+      "class": "Cake\\Network\\Exception\\NotImplementedException",
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: Just curious but would this be better as`"class": \Cake\Network\Exception\NotImplementedException::class` ?

File: docs/listeners/jsonapi.rst [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+JSON API specification.
+
+
+Response formats
+----------------
+
+HTTP GET (index)
+^^^^^^^^^^^^^^^^
+
+Requests to the ``index`` action **must** use:
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: What would the url look like here? `example.com/countries` or `example.com/countries.json` or with `index.json` ?

File: docs/listeners/jsonapi.rst [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+Please note that only support for ``belongsTo`` and ``hasMany``
+relationships has been implemented.
+
+
+Configuration
+-------------
+
+The output produced by the listener is highly configurable using the Crud configuration
+options described in this section. Configure the options on the fly in your controller
+actions or enable them for all actions in your controller by adding them to the
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: 'by adding them to your controllers' I would think here, just an additional 'the'

File: docs/listeners/jsonapi.rst [hunk 6]
```diff
@@ -0,0 +1,723 @@
...
+include
+^^^^^^^
+
+Pass this **array** option (default: empty) an array with associated
+entities to limit the data added to the json ``included`` node.
+
+.. code-block:: phpinline
+
+  $this->Crud->config('listeners.jsonApi.include', [
+    'currency', // belongsTo relations will be singular
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: Do these need to be Capitalised if they're an entity? Or do you mean an entity property?

File: docs/listeners/jsonapi.rst [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+    'cultures' // hasMany relations will be plural
+  ]);
+
+fieldSets
+^^^^^^^^^
+
+Pass this **hash** option (default: empty) a hash with
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: What's a hash? Do you mean an array?

File: docs/listeners/jsonapi.rst [hunk 8]
```diff
@@ -0,0 +1,723 @@
...
+  ]);
+
+fieldSets
+^^^^^^^^^
+
+Pass this **hash** option (default: empty) a hash with 
+field names to limit the attributes/fields shown in the
+generated json.
+
+For exampl, For example
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: Example

File: docs/listeners/jsonapi.rst [hunk 9]
```diff
@@ -0,0 +1,723 @@
...
+This listener makes use of `NeoMerx schemas <https://github.com/neomerx/json-api/wiki/Schemas>`_
+to handle the heavy lifting that is required for converting CakePHP entities to JSON API format.
+
+By default all entities in the ``_entities`` viewVar will be passed to the
+Listener's ``DynamicEntitySchema`` for conversion. This dynamic schema extends
+``Neomerx\JsonApi\Schema\SchemaProvider`` and is, amongst other things, used to
+override NeoMerx methods so we can generate CakePHP specific output (like links).
+
+Even though the dynamic entity schema provided by Crud should cater to the
+needs of most users... creating your own custom schemas is also supported. When
```
Comment (davidyell) [inline docs/listeners/jsonapi.rst]: I'd be tempted to replace the `...` with `, ` here, just for flow.

**Comment by davidyell** [comment]
Great job @bravo-kernel 👍

**Comment by bravo-kernel** [comment]
Thanks for reviewing, I will try to get things fixes tonight 👍🏼

Sent from my Mobile Device

File: docs/listeners/jsonapi.rst [hunk 10]
```diff
@@ -89,9 +89,16 @@ if you prefer attaching the listener to specific controllers on the fly.
 CakeRequest detector
 --------------------
 
-The JsonApi Listener adds the ``is(jsonapi)`` request detector 
-to your ``CakeRequest`` object which check if the request
-contains a ``HTTP Accept`` header  set to ``application/vnd.api+json``.
+The JsonApi Listener adds the ``jsonapi`` request detector
+to your ``CakeRequest`` object which checks if the request
```
Comment (ADmad) [inline docs/listeners/jsonapi.rst]: There's no `CakeRequest` in 3.x. It's just `Request`.

**Comment by bravo-kernel** [comment]
@ADmad changed, for the Api docs as well

**Comment by bravo-kernel** [comment]
sorry guys, it seems forcing removes the pointers to the comments, lesson learned, apologies ☺️

**Comment by ADmad** [comment]
> sorry guys, it seems forcing removes the pointers to the comments, lesson learned, apologies

Doesn't matter, especially not for docs. A cleaner history matters more.

**Comment by bravo-kernel** [comment]
True but I think rebasing after everybody is done reviewing would be more user-friendly/workable.

**Comment by davidyell** [comment]
👍

**Comment by bravo-kernel** [comment]
Thanks for reviewing Mr. @davidyell :shipit:

**Comment by bravo-kernel** [comment]
@jippi if you merge https://github.com/FriendsOfCake/crud/pull/471 this can be merged as well.

**Comment by bravo-kernel** [comment]
Much obliged Mr. @jippi 🍰

**Pull request merged**